gk/fib.c

```diff
@@ -108,19 +169,32 @@ neigh_get_ether_cache_locked(struct neighbor_hash_table *neigh,
 		"Failed to add a cache entry to the neighbor hash table at %s\n",
 		__func__);
 
+	clear_ether_cache(eth_cache);
 	return NULL;
 }
 
 /* Warning: avoid calling this function directly, prefer ether_cache_put(). */
 static int
 neigh_del_ether_cache(struct neighbor_hash_table *neigh, void *key)
 {
-	int ret = rte_hash_del_key(neigh->hash_table, key);
-	if (ret >= 0)
-		return initialize_ether_cache(&neigh->cache_tbl[ret]);
+	struct ether_cache *eth_cache;
+	hash_sig_t hash_val = rte_hash_hash(neigh->hash_table, key);
+	int ret = rte_hash_lookup_with_hash_data(
+		neigh->hash_table, key, hash_val, (void **)&eth_cache);
+	if (ret == 0) {
+		int succ = rte_hash_del_key_with_hash(
+			neigh->hash_table, key, hash_val);
+		if (succ < 0) {
+			RTE_LOG(ERR, HASH,
+				"Failed to delete a cache entry to the neighbor hash table at %s - %s\n",
+				__func__, strerror(-succ));
+		}
+
+		return clear_ether_cache(eth_cache);
```

Comment:
We have a race condition here.

Consider the following. The last fib entry pointing to @eth_cache is going to be removed. After all GK blocks remove their reference to the being-removed fib entry, `ether_cache_put()` is called. Then, `ether_cache_put()` calls `put_arp()`/`put_nd()` and `neigh_del_ether_cache()` aftwards. At this exact moment, the race begins.

As `neigh_del_ether_cache()` is running, it will eventually call `clear_ether_cache()` to clear `eth_cache`. But, concurrently, the LLS block will call `gk_arp_and_nd_req_cb()`, which also edits `eth_cache`.

The easiest solution is to move `clear_ether_cache()` into `gk_arp_and_nd_req_cb()`, so `gk_arp_and_nd_req_cb()` calls `clear_ether_cache()` at its last execution.

Reply:
In fact, a better solution is for `gk_arp_and_nd_req_cb()` to do nothing when it's the last call.

Reply:
I just realized that we must move `clear_ether_cache()` into `gk_arp_and_nd_req_cb()` to be called during the last call of `gk_arp_and_nd_req_cb()`. The reason is that, between the return of `put_arp()`/`put_nd()` and the time `neigh_del_ether_cache()` calls `clear_ether_cache()`, the LLS block may call `gk_arp_and_nd_req_cb()` a number of times before the last call due to network updates.